pandas/core/groupby/grouper.py

```diff
@@ -557,7 +558,16 @@ def indices(self):
             return self.grouper.indices
 
         values = Categorical(self.grouper)
-        return values._reverse_indexer()
+
+        # GH35014
+        reverse_indexer = values._reverse_indexer()
+        if not self.dropna and any(pd.isna(v) for v in values):
```

Comment:
use something more like this (as what you have is non-performant relatively speaking)

```
mask = values.isna()
if not self.dropna and mask.any():
    return {**reverse_indexer, pd.NaT: np.arange(len(values))[mask]}
```

though why are you return pd.NaT here? that is odd